File: leetcode/finished/two_sums.m
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN
@interface Solution : NSObject
+ (NSArray<NSNumber *> *)twoSumWithNums:(NSArray<NSNumber *> *)nums target:(NSInteger)target;
@end
NS_ASSUME_NONNULL_END

@implementation Solution

+ (NSArray<NSNumber *> *)twoSumWithNums:(NSArray<NSNumber *> *)nums target:(NSInteger)target {
    NSMutableDictionary<NSNumber *, NSNumber *> *values = [NSMutableDictionary dictionary];
    for (int i = 0; i < [nums count]; i++) {
        NSNumber *num = nums[i];
        NSInteger diff = target - num.intValue;
        NSNumber *result = values[@(diff)];
        if (result) {
            return @[result, @(i)];
        }
        values[num] = @(i);
    }
    return @[];
}

@end

void test(NSArray<NSNumber *> *nums, NSInteger target) {
    NSLog(@"nums = %@", nums);
    NSArray<NSNumber *> *result = [Solution twoSumWithNums:nums target:target];
    NSLog(@"result = %@", result);
}

int main() {
    test(@[@2, @7, @11, @15], 9); // [0, 1]
    NSLog(@"");
    test(@[@3, @2, @4], 6); // [1, 2]
}

File: leetcode/finished/two_sums.swift
#!/usr/bin/swift

class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
        var values: [Int: Int] = [:];
        for i in 0..<nums.count {
            let num = nums[i]
            let diff = target - num
            if let result = values[diff] {
                return [result, i]
            }
            values[num] = i
        }
        return [];
    }
}

func test(_ nums: [Int], _ target: Int) {
    let solution = Solution();
    print("nums = \(nums)")
    let result = solution.twoSum(nums, target)
    print("result = \(result)")
}

test([2, 7, 11, 15], 9) // [0, 1]
print("")
test([3, 2, 4], 6); // [1, 2]
